File: src/relay_deck/controller.py
from __future__ import annotations

import re
from pathlib import Path

from relay_deck.models import ToolType
from relay_deck.router import RouterResult


class ControllerInterpreter:
    CREATE_KEYWORDS = (
        "create",
        "start",
        "spawn",
        "launch",
        "new agent",
        "new session",
        "创建",
        "新建",
        "启动",
        "开一个",
        "开启",
    )

    NAME_PATTERNS = (
        re.compile(r"\b(?:named|name)\s+(?P<name>[A-Za-z0-9_-]+)\b", re.IGNORECASE),
        re.compile(r"(?:叫|名为)(?P<name>[A-Za-z0-9_-]+)"),
    )

    PATH_PATTERN = re.compile(r"(?P<path>(?:~|/)[^\s,，。；;]+)")

    def interpret(self, text: str) -> RouterResult | None:
        if not text.strip():
            return None
        tool_type = self._detect_tool_type(text)
        cwd = self._detect_cwd(text)
        if tool_type is None or cwd is None or not self._looks_like_create_request(text):
            return None
        name = self._detect_name(text) or self._derive_name(tool_type, cwd)
        return RouterResult(
            kind="create_agent",
            tool_type=tool_type,
            name=name,
            cwd=cwd,
            message=text,
        )

    def _looks_like_create_request(self, text: str) -> bool:
        lowered = text.lower()
        return any(keyword in lowered for keyword in self.CREATE_KEYWORDS)

    def _detect_tool_type(self, text: str) -> ToolType | None:
        lowered = text.lower()
        if "claude code" in lowered or "claude" in lowered:
            return ToolType.CLAUDE
        if "codex" in lowered:
            return ToolType.CODEX
        return None

    def _detect_name(self, text: str) -> str | None:
        for pattern in self.NAME_PATTERNS:
            match = pattern.search(text)
            if match:
                return match.group("name")
        return None

    def _detect_cwd(self, text: str) -> Path | None:
        match = self.PATH_PATTERN.search(text)
        if match is None:
            return None
        return Path(match.group("path")).expanduser()

    def _derive_name(self, tool_type: ToolType, cwd: Path) -> str:
        base = cwd.name or "workspace"
        normalized = re.sub(r"[^A-Za-z0-9_-]+", "-", base).strip("-").lower() or "workspace"
        return f"{normalized}-{tool_type.value}"

File: src/lib/controller.ts
import { RouterResult, type AgentTool } from "./types";

const CREATE_KEYWORDS = [
  "create",
  "start",
  "spawn",
  "launch",
  "new agent",
  "new session",
  "创建",
  "新建",
  "启动",
  "开一个",
  "开启",
];

const NAME_PATTERNS = [
  /\b(?:named|name)\s+(?<name>[A-Za-z0-9_-]+)\b/i,
  /(?:叫|名为)(?<name>[A-Za-z0-9_-]+)/,
];

const PATH_PATTERN = /(?<path>(?:~|\/)[^\s,，。；;]+)/;

function detectToolType(text: string): AgentTool | null {
  const lowered = text.toLowerCase();
  if (lowered.includes("claude code") || lowered.includes("claude")) {
    return "claude";
  }
  if (lowered.includes("codex")) {
    return "codex";
  }
  return null;
}

function detectName(text: string): string | null {
  for (const pattern of NAME_PATTERNS) {
    const match = pattern.exec(text);
    if (match?.groups?.name) {
      return match.groups.name;
    }
  }
  return null;
}

function detectPath(text: string): string | null {
  const match = PATH_PATTERN.exec(text);
  return match?.groups?.path ?? null;
}

function deriveName(tool: AgentTool, cwd: string): string {
  const base = cwd.split("/").filter(Boolean).pop() || "workspace";
  const normalized = base.replace(/[^A-Za-z0-9_-]+/g, "-").replace(/^-+|-+$/g, "").toLowerCase() || "workspace";
  return `${normalized}-${tool}`;
}

export function interpretControllerMessage(text: string): RouterResult | null {
  if (!text.trim()) {
    return null;
  }
  const lowered = text.toLowerCase();
  if (!CREATE_KEYWORDS.some((keyword) => lowered.includes(keyword))) {
    return null;
  }
  const tool = detectToolType(text);
  const cwd = detectPath(text);
  if (!tool || !cwd) {
    return null;
  }

  return {
    kind: "create_agent",
    tool,
    cwd,
    name: detectName(text) || deriveName(tool, cwd),
    message: text,
  };
}
